From 3062d8a3ea620b040487ba06ac0f61e678e4d04f Mon Sep 17 00:00:00 2001
From: Antony Abee Prakash XV <antonyabee.prakashxv@adlinktech.com>
Date: Fri, 3 Jul 2020 12:48:52 +0530
Subject: [PATCH 15/15] Add-OV5640-mipi_csi-camera-support

This patch will modify the pinmux for csi1 and csi2 and add ov5640 2lane camera
support in i2c4 bus. Also removes csi2 and mipi_csi_2 nodes from devicetree as
OV5640 2lane camera is connected on csi1

Signed-off-by: Antony Abee Prakash XV <antonyabee.prakashxv@adlinktech.com>
---
 .../boot/dts/adlink/adlink-lec-imx8m.dts      | 128 ++++++------------
 1 file changed, 39 insertions(+), 89 deletions(-)

diff --git a/arch/arm64/boot/dts/adlink/adlink-lec-imx8m.dts b/arch/arm64/boot/dts/adlink/adlink-lec-imx8m.dts
index a2b1192db927..8a020a6c1a19 100755
--- a/arch/arm64/boot/dts/adlink/adlink-lec-imx8m.dts
+++ b/arch/arm64/boot/dts/adlink/adlink-lec-imx8m.dts
@@ -9,6 +9,7 @@
 #include <dt-bindings/usb/pd.h>
 #include "imx8mq.dtsi"
 #include <dt-bindings/net/ti-dp83867.h>
+#define csien 1
 #define sdslow 1
 
 / {
@@ -174,6 +175,7 @@
 	cpu-supply = <&buck2_reg>;
 };
 
+#if csien
 &csi1_bridge {
 	fsl,mipi-mode;
 	fsl,two-8bit-sensor-mode;
@@ -186,17 +188,23 @@
 	};
 };
 
-&csi2_bridge {
-	fsl,mipi-mode;
-	fsl,two-8bit-sensor-mode;
+&mipi_csi_1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
 	status = "okay";
-
 	port {
-		csi2_ep: endpoint {
-			remote-endpoint = <&csi2_mipi_ep>;
+		mipi1_sensor_ep: endpoint@0 {
+			remote-endpoint = <&ov5640_mipi1_ep>;
+			data-lanes = <1 2>;
+			bus-type = <4>;
+		};
+
+		csi1_mipi_ep: endpoint@1 {
+			remote-endpoint = <&csi1_ep>;
 		};
 	};
 };
+#endif
 
 &fec1 {
 	pinctrl-names = "default";
@@ -225,40 +233,6 @@
 	};
 };
 
-&mipi_csi_1 {
-	#address-cells = <1>;
-	#size-cells = <0>;
-	status = "okay";
-	port {
-		mipi1_sensor_ep: endpoint@0 {
-			remote-endpoint = <&ov5640_mipi1_ep>;
-			data-lanes = <1 2>;
-			bus-type = <4>;
-		};
-
-		csi1_mipi_ep: endpoint@1 {
-			remote-endpoint = <&csi1_ep>;
-		};
-	};
-};
-
-&mipi_csi_2 {
-	#address-cells = <1>;
-	#size-cells = <0>;
-	status = "okay";
-	port {
-		mipi2_sensor_ep: endpoint@0 {
-			remote-endpoint = <&ov5640_mipi2_ep>;
-			data-lanes = <1 2>;
-			bus-type = <4>;
-		};
-
-		csi2_mipi_ep: endpoint@1 {
-			remote-endpoint = <&csi2_ep>;
-		};
-	};
-};
-
 &sai2 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_sai2>;
@@ -286,28 +260,6 @@
 	pinctrl-0 = <&pinctrl_i2c1>;
 	status = "okay";
 
-	ov5640_mipi2: ov5640_mipi2@3c {
-		compatible = "ovti,ov5640_mipi";
-		reg = <0x3c>;
-		status = "okay";
-		pinctrl-names = "default";
-		pinctrl-0 = <&pinctrl_csi2_pwn>, <&pinctrl_csi_rst>;
-		clocks = <&clk IMX8MQ_CLK_CLKO2>;
-		clock-names = "csi_mclk";
-		assigned-clocks = <&clk IMX8MQ_CLK_CLKO2>;
-		assigned-clock-parents = <&clk IMX8MQ_SYS2_PLL_200M>;
-		assigned-clock-rates = <20000000>;
-		csi_id = <1>;
-		pwn-gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>;
-		mclk = <20000000>;
-		mclk_source = <0>;
-		port {
-			ov5640_mipi2_ep: endpoint {
-				remote-endpoint = <&mipi2_sensor_ep>;
-			};
-		};
-	};
-
 	pmic@8 {
 		compatible = "fsl,pfuze100";
 		fsl,pfuze-support-disable-sw;
@@ -423,12 +375,31 @@
 	pinctrl-0 = <&pinctrl_i2c2>;
 	status = "okay";
 
+        sgtl5000: codec@2a {
+                compatible = "fsl,sgtl5000";
+                reg = <0x2a>;
+                clocks = <&clk IMX8MQ_CLK_SAI2_ROOT>;
+                clock-names = "mclk";
+                //micbias-resistor-k-ohms       = <2>;
+                //micbias-voltage-m-volts       = <2250>;
+                //VDDA-supply                           = <&reg_XXX>;
+                //VDDIO-supply                          = <&reg_XXX>;
+                //VDDD-supply                           = <&reg_XXX>;
+        };
+};
+
+&i2c4 {
+	clock-frequency = <100000>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c4>;
+	status = "okay";
+
         ov5640_mipi: ov5640_mipi@3c {
                 compatible = "ovti,ov5640_mipi";
                 reg = <0x3c>;
                 status = "okay";
                 pinctrl-names = "default";
-                pinctrl-0 = <&pinctrl_csi1_pwn>;
+                pinctrl-0 = <&pinctrl_csi1>;
                 clocks = <&clk IMX8MQ_CLK_CLKO2>;
                 clock-names = "csi_mclk";
                 assigned-clocks = <&clk IMX8MQ_CLK_CLKO2>;
@@ -436,6 +407,7 @@
                 assigned-clock-rates = <20000000>;
                 csi_id = <0>;
                 pwn-gpios = <&gpio1 3 GPIO_ACTIVE_HIGH>;
+		rst-gpios = <&gpio1 6 GPIO_ACTIVE_HIGH>;
                 mclk = <20000000>;
                 mclk_source = <0>;
                 port {
@@ -444,25 +416,6 @@
                         };
                 };
         };
-
-        sgtl5000: codec@2a {
-                compatible = "fsl,sgtl5000";
-                reg = <0x2a>;
-                clocks = <&clk IMX8MQ_CLK_SAI2_ROOT>;
-                clock-names = "mclk";
-                //micbias-resistor-k-ohms       = <2>;
-                //micbias-voltage-m-volts       = <2250>;
-                //VDDA-supply                           = <&reg_XXX>;
-                //VDDIO-supply                          = <&reg_XXX>;
-                //VDDD-supply                           = <&reg_XXX>;
-        };
-};
-
-&i2c4 {
-	clock-frequency = <100000>;
-	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_i2c4>;
-	status = "okay";
 };
 
 &pcie0 {
@@ -594,19 +547,16 @@
 
 	};
 
-	pinctrl_csi1_pwn: csi1_pwn_grp {
+	pinctrl_csi1: csi1_pwn_grp {
 		fsl,pins = <
 			MX8MQ_IOMUXC_GPIO1_IO03_GPIO1_IO3		0x19
+			MX8MQ_IOMUXC_GPIO1_IO06_GPIO1_IO6		0x19
+			MX8MQ_IOMUXC_GPIO1_IO15_CCMSRCGPCMIX_CLKO2	0x59
 		>;
 	};
-	pinctrl_csi2_pwn: csi2_pwn_grp {
+	pinctrl_csi2: csi2_pwn_grp {
 		fsl,pins = <
 			MX8MQ_IOMUXC_GPIO1_IO05_GPIO1_IO5		0x19
-		>;
-	};
-
-	pinctrl_csi_rst: csi_rst_grp {
-		fsl,pins = <
 			MX8MQ_IOMUXC_GPIO1_IO06_GPIO1_IO6		0x19
 			MX8MQ_IOMUXC_GPIO1_IO15_CCMSRCGPCMIX_CLKO2	0x59
 		>;
-- 
2.17.1

